kart/diff_estimation.py

```diff
@@ -197,24 +206,39 @@ def _nonrecursive_diff(tree_a, tree_b):
     return {k: (a.get(k), b.get(k)) for k in all_names if a.get(k) != b.get(k)}
 
 
-def _recursive_distributed_diff_count(
+def _num_expected_distributed_tree_blobs(num_samples, branch_factor):
+    """
+    Returns the expected number of children in a tree of the given size.
+
+    """
+    # https://docs.google.com/document/d/11CeJKbiNQoLmhDcYIM68cJSA_nKBHW7kYVybh2N-Lww/edit#heading=h.7z95y6hc62gn
+    return math.log(1 - num_samples / branch_factor) / math.log(1 - 1 / branch_factor)
+
+
+def _recursive_distributed_diff_estimate(
     repo, tree1, tree2, branch_count, total_samples_to_take
 ):
     diff = _nonrecursive_diff(tree1, tree2)
+
     diff_size = len(diff)
     if diff_size < branch_count / 2:
-        L.debug(f"Found {diff_size} diffs.")
-        return diff_size, 1
+        estimated_blobs = _num_expected_distributed_tree_blobs(diff_size, branch_count)
+        L.debug(f"Found {diff_size} diffs for an estimate of {estimated_blobs} blobs.")
+        return estimated_blobs, 1
 
     L.debug(f"Found {diff_size} diffs, checking next level:")
 
     total_subsample_size = 0
     total_subsamples_taken = 0
     total_samples_taken = 0
     for tree1, tree2 in diff.values():
-        subsample_size, samples_taken = _recursive_distributed_diff_count(
-            repo, tree1, tree2, branch_count, total_samples_to_take
-        )
+        if isinstance(tree1, pygit2.Blob) or isinstance(tree2, pygit2.Blob):
```

Comment:
FYI instead of checking the type of the object, you can access its `type_str` attribute which will be either "blob" or "tree". Makes no difference here so I don't care which you use - but it can make tests easier to write since you can create test objects that present the same interface as pygit2.Blob or pygit2.Tree, without actually using pygit2 at all. So, just keep in mind if you need it

Reply:
sounds good. Here though, the trees might also be None, so I'd have to guard against that if I started accessing attributes on them, whereas the `isinstance` check works regardless